functorch/_src/aot_autograd.py

```diff
@@ -312,25 +312,31 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Tensor], aot_config: AOTConfi
     # else is just getting the types right.
 
     seen_args = {}
-    deduped_flat_args = []
-    drop_args = set()
+    keep_arg_mask = []
+    dropped_args = False
     add_dupe_map = {}
     duped_arg_len = len(flat_args)
+
+    j = 0  # index into deduped_flat_args
     for i, t in enumerate(flat_args):
         if t in seen_args:
-            drop_args.add(i)
+            keep_arg_mask.append(False)
+            dropped_args = True
             add_dupe_map[i] = seen_args[t]
             continue
-        seen_args[t] = len(deduped_flat_args)
-        add_dupe_map[i] = len(deduped_flat_args)
-        deduped_flat_args.append(t)
+        keep_arg_mask.append(True)
+        seen_args[t] = j
+        add_dupe_map[i] = j
+        j += 1
 
+    # NB: Hot path, avoid set lookups here
     def remove_dupe_args(args):
+        if not dropped_args:
+            return args
         r = []
-        for i, t in enumerate(args):
-            if i in drop_args:
-                continue
-            r.append(t)
+        for t, keep in zip(args, keep_arg_mask):
```

Comment:
is simpler as a list comprehension? `return [t for t, keep in zip(args, keep_arg_mask) if keep]`

same for a function below?

Reply:
sure